src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTests.scala

```diff
@@ -404,6 +404,44 @@ class E2EHyperspaceRulesTests extends HyperspaceSuite with SQLHelper {
     assert(!spark.isHyperspaceEnabled())
   }
 
+  test("Validate index usage after refresh with some source data file deleted.") {
+    withSQLConf(
+      IndexConstants.INDEX_LINEAGE_ENABLED -> "true",
+      IndexConstants.REFRESH_DELETE_ENABLED -> "true") {
+
+      // Save a copy of source data files.
+      val location = testDir + "ixRefreshTest"
+      val dataPath = new Path(location, "*parquet")
+      val dataColumns = Seq("c1", "c2", "c3", "c4", "c5")
+      SampleData.save(spark, location, dataColumns)
+
+      // Create index on original source data files.
+      val df = spark.read.parquet(location)
+      val indexConfig = IndexConfig("filterIndex", Seq("c3"), Seq("c1"))
+      hyperspace.createIndex(df, indexConfig)
+
+      // Delete some source data file.
+      val dataFileNames = dataPath
+        .getFileSystem(new Configuration)
+        .globStatus(dataPath)
+        .map(_.getPath)
+
+      assert(dataFileNames.length > 0)
+      val fileToDelete = dataFileNames.head
+      FileUtils.delete(fileToDelete)
+
+      // Refresh the index to remove deleted source data file records from index.
+      hyperspace.refreshIndex(indexConfig.indexName)
+
+      def query(): DataFrame =
+        spark.read.parquet(location).filter("c3 == 'facebook'").select("c3", "c1")
+
+      // Verify index usage on latest version of index (v=1) after refresh.
```

Comment:
I modified the test case; However the reason those checks (usage with v=0 and no usage before refresh) were not added in the first place was because they are being tested by other test case. If an index is not leveraged right after creation (v=0 case); then some other test case will capture. Same is true for no index usage after source data file changes.
I think keeping tests simple and avoiding validating a scenario multiple times in different tests is the recommended practice; unless there is sth else here that I am missing.

Reply:
Well, if there is no verification before applying refresh, we are not really validating anything (it's possible that the test set up was wrong, refresh didn't work, etc.). To me, it's crucial to test this e2e scenario in E2E tests.